Combined patch:

bz 411171


Commit:     3cc3eb1148e4b2dfabf7a1dcf36fd8be1331ca95
Parent:     b90fe23bd51c6b1c298159591c833bdd24f55002
Author:     Jeff Garzik <jeff@garzik.org>
AuthorDate: Wed Sep 26 00:02:41 2007 -0400
Committer:  Jeff Garzik <jeff@garzik.org>
CommitDate: Fri Oct 12 14:55:42 2007 -0400

    [libata] AHCI: enable AHCI mode, before using AHCI reset


Commit:     ab6fc95f609b372a19e18ea689986846ab1ba29c
Parent:     360737a982b1ae09e1659e0bb27085c03f02f404
Author:     Jeff Garzik <jeff@garzik.org>
AuthorDate: Mon Oct 29 10:43:55 2007 -0400
Committer:  Jeff Garzik <jeff@garzik.org>
CommitDate: Mon Oct 29 10:43:55 2007 -0400

    [libata] AHCI: fix newly introduced host-reset bug

---
 drivers/ata/ahci.c |   10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

--- linux-2.6.23.noarch.orig/drivers/ata/ahci.c
+++ linux-2.6.23.noarch/drivers/ata/ahci.c
@@ -872,8 +872,16 @@ static int ahci_reset_controller(struct 
 	void __iomem *mmio = host->iomap[AHCI_PCI_BAR];
 	u32 tmp;
 
-	/* global controller reset */
+	/* we must be in AHCI mode, before using anything
+	 * AHCI-specific, such as HOST_RESET.
+	 */
 	tmp = readl(mmio + HOST_CTL);
+	if (!(tmp & HOST_AHCI_EN)) {
+		tmp |= HOST_AHCI_EN;
+		writel(tmp, mmio + HOST_CTL);
+	}
+
+	/* global controller reset */
 	if ((tmp & HOST_RESET) == 0) {
 		writel(tmp | HOST_RESET, mmio + HOST_CTL);
 		readl(mmio + HOST_CTL); /* flush */
